src/DependencyInjection/CompilerPass/SubscriberPass.php

```diff
@@ -39,4 +46,117 @@ public function process(ContainerBuilder $container)
+     *     </service>
+     *
+     * to the  following code in the container (rough equivalent)
+     *
+     *     public function getFoo()
+     *     {
+     *         $client = new GuzzleHttp\Client();
+     *         $client->getEmitter()->attach($this->get('foo'));
+     *         $client->getEmitter()->attach($this->get('bar'));
+     *
+     *         return $client;
+     *     }
+     *
+     * @param ContainerBuilder $container
+     */
+    private function addSubscribersToClients(ContainerBuilder $container)
+    {
+        $taggedSubscriberIds = $container->findTaggedServiceIds(self::SUBSCRIBER_TAG);
+        $subscriberIds = $this->mapSubscriberIdsByAlias($taggedSubscriberIds);
+
+        foreach ($this->findSubscriberAliasesByClientId($container) as $clientId => $aliases) {
+            $client = $container->findDefinition($clientId);
+
+            $configurator = $this->createConfigurator(array_values($aliases
+                ? array_intersect_key($subscriberIds, array_flip($aliases))
+                : $subscriberIds
+            ));
+
+            // Wraps the previous configurator in case the client already had one.
+            $configurator->addArgument($client->getConfigurator());
+
+            $configuratorId = sprintf('csa_guzzle._configurator.%s', $clientId);
+            $container->setDefinition($configuratorId, $configurator);
+
+            $client->setConfigurator([new Reference($configuratorId), 'configure']);
+        }
+    }
+
+    /**
+     * Finds all tagged clients and lists their registered subscribers by client
+     * ids. Empty arrays are returned for clients specifying no subscribers.
+     *
+     * @param ContainerBuilder $container
+     *
+     * @return array An array of subscriber ids with their consuming client as a key
+     */
+    private function findSubscriberAliasesByClientId(ContainerBuilder $container)
+    {
+        $tagsByClientId = $container->findTaggedServiceIds(self::CLIENT_TAG);
+        $aliases = [];
+
+        foreach ($tagsByClientId as $clientId => $tags) {
+            $subscribers = [];
+
+            foreach ($tags as $tag) {
+                if (isset($tag['subscribers'])) {
+                    $subscribers = array_merge(
+                        $subscribers,
+                        array_map('trim', explode(',', $tag['subscribers']))
+                    );
+                }
+            }
+
+            $aliases[$clientId] = $subscribers;
+        }
+
+        return $aliases;
+    }
+
+    /**
+     * @param array $subscriberIds List of service ids pointing to subscribers
+     *
+     * @return Definition
+     */
+    private function createConfigurator(array $subscriberIds)
+    {
+        $subscriberRefs = array_map(function ($subscriberId) {
+            return new Reference($subscriberId);
+        }, $subscriberIds);
+
+        $configurator = new Definition('Csa\Bundle\GuzzleBundle\DependencyInjection\Configurator\ClientConfigurator');
```

Comment:
Instead of creating a new configurator here, I would actually define an abstract, private service in the XML configuration with null constructor arguments. The client's configurator would then be a DefinitionDecorator instead, and we would use `replaceArgument` instead of `addArgument`.